lib/Partitioner/Partitioner.cpp

```diff
@@ -423,6 +424,31 @@ NodeToFunctionMap Partitioner::selectPartitions(Function *F,
 /// there is only 2 devices.
 void Partitioner::adjustLogicalDeviceID(DAGNode *DAG, int num) {}
 
+/// Duplicate the network to saturate the number of devices. If a network is
+/// partitioned into two parts and there are six devices this would duplicate
+/// the network three times.
+void Partitioner::saturateHost(unsigned logicalDeviceCount) {
+  unsigned duplications = deviceInfo_.size() / logicalDeviceCount;
+  if (duplications < 2) {
+    return;
+  }
+  // Add additional logical devices to each node.
+  for (auto &network : partitions_) {
+    for (auto &node : network.nodes) {
+      // Build list of new logical devices to add to node.
+      std::vector<unsigned> newDevices;
+      for (auto logical : node->logicalDevices) {
+        for (unsigned i = 1; i < duplications; i++) {
```

Comment:
can you add some comments on this logic, like why it starts at 1 and what logical + (i * logicalDeviceCount) is for